File: task/src/main/java/com/decena/task/Service/TaskService.java
package com.decena.task.Service;

import java.util.List;

import com.decena.task.Dto.TaskRequest;
import com.decena.task.Dto.TaskResponse;
import com.decena.task.Dto.TaskStatusUpdateRequest;
import com.decena.task.Dto.TaskBoardReorderRequest;

public interface TaskService {
    TaskResponse createTask(TaskRequest request);
    TaskResponse getTaskById(Long id);
    List<TaskResponse> getAllTasks(int page, int size, String sortBy, String sortDir);
    TaskResponse updateTask(Long id, TaskRequest request);
    void deleteTask(Long id);
    TaskResponse markTaskAsCompleted(Long id);
    List<TaskResponse> getDeletedTasks(int page, int size);

    /**
     * Restores a soft-deleted task.
     *
     * @param id task ID
     * @return restored task response
     * @throws com.decena.task.Exception.ResourceNotFoundException when task does not exist
     * @throws IllegalArgumentException when task is not deleted
     */
    TaskResponse restoreTask(Long id);

    /**
     * Updates the Kanban status of one active task.
     *
     * @param id task ID
     * @param request status payload
     * @return updated task response
     * @throws com.decena.task.Exception.ResourceNotFoundException when task does not exist
     * @throws IllegalArgumentException when status is invalid
     */
    TaskResponse updateTaskStatus(Long id, TaskStatusUpdateRequest request);

    /**
     * Reorders tasks inside one Kanban column.
     *
     * @param request reorder payload
     * @throws IllegalArgumentException when payload is invalid
     */
    void reorderBoard(TaskBoardReorderRequest request);

}

File: frontend/src/services/taskService.ts
import axios from 'axios';
import { TaskRequest, TaskResponse } from '../types/task';

const API_URL = 'http://localhost:8080/api/tasks';

/**
 * Get all active tasks
 */
export const getAllTasks = async (): Promise<TaskResponse[]> => {
  const res = await axios.get<TaskResponse[]>(`${API_URL}`);
  return res.data;
};

/**
 * Get a single task by ID
 */
export const getTaskById = async (id: number): Promise<TaskResponse> => {
  const res = await axios.get<TaskResponse>(`${API_URL}/${id}`);
  return res.data;
};

/**
 * Create a new task
 */
export const createTask = async (task: TaskRequest): Promise<TaskResponse> => {
  const res = await axios.post<TaskResponse>(`${API_URL}`, task);
  return res.data;
};

/**
 * Update an existing task
 */
export const updateTask = async (id: number, task: TaskRequest): Promise<TaskResponse> => {
  const res = await axios.put<TaskResponse>(`${API_URL}/${id}`, task);
  return res.data;
};

/**
 * Soft delete a task
 */
export const deleteTask = async (id: number): Promise<void> => {
  await axios.delete(`${API_URL}/${id}`);
};

/**
 * Mark a task as completed
 */
export const markTaskAsCompleted = async (id: number): Promise<TaskResponse> => {
  const res = await axios.patch<TaskResponse>(`${API_URL}/${id}/complete`);
  return res.data;
};

/**
 * Get deleted tasks
 */
export const getDeletedTasks = async (): Promise<TaskResponse[]> => {
  const res = await axios.get<TaskResponse[]>(`${API_URL}/deleted`);
  return res.data;
};
